File: projectpastibisa/TransferActivity.java
package com.example.projectpastibisa;

import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.database.*;

public class TransferActivity extends AppCompatActivity {

    private EditText etRecipientPhone, etAmount;
    private Button btnTransfer;

    private FirebaseAuth mAuth;
    private DatabaseReference database;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_transfer);

        etRecipientPhone = findViewById(R.id.etRecipientPhone);
        etAmount = findViewById(R.id.etAmount);
        btnTransfer = findViewById(R.id.btnTransfer);

        mAuth = FirebaseAuth.getInstance();
        database = FirebaseDatabase.getInstance().getReference();

        // Ambil nomor telepon dari intent dan set ke etRecipientPhone
        String phoneNumber = getIntent().getStringExtra("nomor_telepon");
        if (phoneNumber != null) {
            etRecipientPhone.setText(phoneNumber);
        }

        btnTransfer.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                String recipientPhone = etRecipientPhone.getText().toString();
                String amountStr = etAmount.getText().toString();

                if (recipientPhone.isEmpty() || amountStr.isEmpty()) {
                    Toast.makeText(getApplicationContext(), "Semua kolom harus diisi", Toast.LENGTH_SHORT).show();
                } else {
                    int amount = Integer.parseInt(amountStr);
                    transferMoney(recipientPhone, amount);
                }
            }
        });
    }

    private void transferMoney(String recipientPhone, int amount) {
        String senderId = mAuth.getCurrentUser().getUid();

        database.child("users").child(senderId).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    Integer senderBalanceObj = dataSnapshot.child("balance").getValue(Integer.class);
                    if (senderBalanceObj != null) {
                        int senderBalance = senderBalanceObj;

                        if (senderBalance >= amount) {
                            DatabaseReference recipientRef = database.child("users");
                            recipientRef.orderByChild("phone_number").equalTo(recipientPhone).addListenerForSingleValueEvent(new ValueEventListener() {
                                @Override
                                public void onDataChange(@NonNull DataSnapshot recipientSnapshot) {
                                    if (recipientSnapshot.exists()) {
                                        String recipientId = recipientSnapshot.getChildren().iterator().next().getKey();
                                        Integer recipientBalanceObj = recipientSnapshot.child(recipientId).child("balance").getValue(Integer.class);
                                        if (recipientBalanceObj != null) {
                                            int recipientBalance = recipientBalanceObj;

                                            // Update sender and recipient balances
                                            database.child("users").child(senderId).child("balance").setValue(senderBalance - amount);
                                            database.child("users").child(recipientId).child("balance").setValue(recipientBalance + amount);

                                            // Save transaction record
                                            DatabaseReference transactionRef = database.child("transactions").push();
                                            transactionRef.child("from").setValue(senderId);
                                            transactionRef.child("to").setValue(recipientId);
                                            transactionRef.child("amount").setValue(amount);
                                            transactionRef.child("timestamp").setValue(ServerValue.TIMESTAMP);

                                            Toast.makeText(getApplicationContext(), "Transfer berhasil", Toast.LENGTH_SHORT).show();
                                        } else {
                                            Toast.makeText(getApplicationContext(), "Saldo penerima tidak ditemukan", Toast.LENGTH_SHORT).show();
                                        }
                                    } else {
                                        Toast.makeText(getApplicationContext(), "Nomor penerima tidak ditemukan", Toast.LENGTH_SHORT).show();
                                    }
                                }

                                @Override
                                public void onCancelled(@NonNull DatabaseError databaseError) {
                                    Log.e("FirebaseDB", "Gagal mencari nomor penerima", databaseError.toException());
                                }
                            });
                        } else {
                            Toast.makeText(getApplicationContext(), "Saldo tidak cukup", Toast.LENGTH_SHORT).show();
                        }
                    } else {
                        Toast.makeText(getApplicationContext(), "Saldo pengirim tidak ditemukan", Toast.LENGTH_SHORT).show();
                    }
                } else {
                    Toast.makeText(getApplicationContext(), "Pengirim tidak ditemukan", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {
                Log.e("FirebaseDB", "Gagal mencari saldo pengirim", databaseError.toException());
            }
        });
    }
}

File: app/src/main/java/com/example/pinkiewallet/backend/TransferActivity.kt
package com.example.pinkiewallet.backend;

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.example.pinkiewallet.R
import com.example.pinkiewallet.backend.Payment
import com.example.pinkiewallet.backend.PinReqActivity
import com.example.pinkiewallet.databinding.TransferBinding
import com.example.pinkiewallet.viewmodel.TransferViewModel
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener

class TransferActivity : AppCompatActivity() {

    private lateinit var binding: TransferBinding

    private lateinit var mAuth: FirebaseAuth
    private lateinit var database: FirebaseDatabase
    private lateinit var usersRef: DatabaseReference
    private var balanceListener: ValueEventListener? = null

    private val transferViewModel: TransferViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = TransferBinding.inflate(layoutInflater) // Updated to use the new layout file name
        setContentView(binding.root)

        mAuth = FirebaseAuth.getInstance()
        database = FirebaseDatabase.getInstance()
        usersRef = database.getReference("users")

        setupUI()
        initFirebase()

        // Hide BottomNavigationView
        val bottomNavigationView = findViewById<BottomNavigationView>(R.id.nav_view)
        bottomNavigationView.visibility = View.GONE

        binding.backwardButton.setOnClickListener {
            finish()
            bottomNavigationView.visibility = View.VISIBLE
        }

        val phoneNumber = intent.getStringExtra("nomor_telepon")
        if (phoneNumber != null) {
            binding.nomorPenerima.setText(phoneNumber)
        }

        binding.btnContinue.setOnClickListener {
            val recipientPhone = binding.nomorPenerima.text.toString()
            val amountStr = binding.etTransferAmount.text.toString()

            if (recipientPhone.isEmpty() || amountStr.isEmpty()) {
                Toast.makeText(this, "Semua kolom harus diisi", Toast.LENGTH_SHORT).show()
            } else {
                val amount = amountStr
                initiatePinRequest(recipientPhone, amount)
            }
        }

        transferViewModel.transferResult.observe(this, Observer { transferSuccessful ->
            if (transferSuccessful) {
                Toast.makeText(this, "Transfer berhasil - Ke Payment", Toast.LENGTH_SHORT).show()
                val intent = Intent(this, Payment::class.java)
                intent.putExtra("jumlah_harga", binding.etTransferAmount.text.toString())
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                startActivity(intent)
            } else {
                Toast.makeText(this, "Transfer gagal", Toast.LENGTH_SHORT).show()
            }
        })
    }

    private fun initFirebase() {
        val userId = mAuth.currentUser?.uid
        if (userId != null) {
            balanceListener = createBalanceListener()
            usersRef.child(userId).addValueEventListener(balanceListener!!)
        } else {
            Toast.makeText(this, "User not authenticated", Toast.LENGTH_SHORT).show()
        }
    }

    private fun createBalanceListener(): ValueEventListener {
        return object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    val balance = dataSnapshot.child("balance").getValue(Long::class.java)
                    val formattedBalance = formatToRupiah(balance ?: 0)
                    binding.saldodapatditransfer.text = "Saldo yang Dapat Ditransfer: $formattedBalance"
                } else {
                }
            }

            override fun onCancelled(databaseError: DatabaseError) {
                Log.e("FirebaseDB", "Failed to retrieve balance", databaseError.toException())
            }
        }
    }

    private fun setupUI() {
        binding.etTransferAmount.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable?) {
                val amountString = s.toString()
                if (amountString.isNotEmpty()) {
                    val amount = amountString.toLong()
                    val formattedAmount = formatToRupiah(amount)
                    binding.jumlahpembayaran.text = formattedAmount
                    binding.btnContinue.setBackgroundColor(ContextCompat.getColor(this@TransferActivity, R.color.teal_200))
                } else {
                    binding.jumlahpembayaran.text = "Rp0"
                    binding.btnContinue.setBackgroundColor(ContextCompat.getColor(this@TransferActivity, android.R.color.darker_gray))
                }
            }
        })

        binding.btnContinue.setBackgroundColor(ContextCompat.getColor(this, android.R.color.darker_gray))
    }

    private fun formatToRupiah(amount: Long): String {
        val formatter = java.text.NumberFormat.getCurrencyInstance(java.util.Locale("id", "ID"))
        return formatter.format(amount).replace("Rp", "Rp ")
    }

    private fun initiatePinRequest(nomorHP: String, jumlahHarga: String) {
        val pinReqIntent = Intent(this, PinReqActivity::class.java)
        pinReqIntent.putExtra("caller", "TransferActivity")
        pinReqIntent.putExtra("jumlah_harga", jumlahHarga)
        pinReqIntent.putExtra("nomorHp", nomorHP)
        startActivity(pinReqIntent)
    }

    override fun onDestroy() {
        super.onDestroy()
        balanceListener?.let {
            mAuth.currentUser?.uid?.let { userId ->
                usersRef.child(userId).removeEventListener(it)
            }
        }
    }
}
